File: library/src/main/java/me/shkschneider/skeleton/security/HmacHelper.java
package me.shkschneider.skeleton.security;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import me.shkschneider.skeleton.helper.LogHelper;
import me.shkschneider.skeleton.java.StringHelper;

public class HmacHelper {

    private static final String ALGORITHM = "HmacSHA1";

    @Nullable
    public static String hash(@NonNull final String key, @NonNull final String string) {
        try {
            final SecretKeySpec secretKeySpec = new SecretKeySpec(key.getBytes(), ALGORITHM);
            final Mac mac = Mac.getInstance(ALGORITHM);
            mac.init(secretKeySpec);
            final byte[] digest = mac.doFinal(string.getBytes());
            if (digest == null) {
                return null;
            }
            return StringHelper.hexadecimal(digest);
        }
        catch (final NoSuchAlgorithmException e) {
            LogHelper.wtf(e);
            return null;
        }
        catch (final InvalidKeyException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

}

File: library/src/main/java/me/shkschneider/skeleton/security/HmacHelper.kt
package me.shkschneider.skeleton.security

import me.shkschneider.skeleton.helperx.Logger
import me.shkschneider.skeleton.java.StringHelper
import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object HmacHelper {

    private const val ALGORITHM = "HmacSHA1"

    fun hash(key: String, string: String): String? {
        try {
            val secretKeySpec = SecretKeySpec(key.toByteArray(), ALGORITHM)
            val mac = Mac.getInstance(ALGORITHM)
            mac.init(secretKeySpec)
            val digest = mac.doFinal(string.toByteArray()) ?: return null
            return StringHelper.hexadecimal(digest)
        } catch (e: NoSuchAlgorithmException) {
            Logger.wtf(e)
            return null
        } catch (e: InvalidKeyException) {
            Logger.wtf(e)
            return null
        }
    }

    fun algorithm(): String {
        return ALGORITHM
    }

}
